include/fs/fs_sys.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_FS_FS_SYS_H_
+#define ZEPHYR_INCLUDE_FS_FS_SYS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @ingroup file_system_api
+ * @{
+ */
+
+#define FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION(type) 			\
+	static type file_system_unsupported_operation_##type(void)	\
+	{								\
+		return -ENOTSUP;					\
+	}
+
+FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION(off_t);
+FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION(ssize_t);
+FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION(int);
+
+#define FILE_SYSTEM_UNSUPPORTED_OPERATION(type)				\
+	(void *)file_system_unsupported_operation_##type
+
+/**
+ * @brief Define file_system_t API callback table for file system
+ *
+ * The macro defines static file_system_t object, and assigns provided
+ * callback functions to operations; operations that are not assigned
+ * any function will be assigned default function returning -ENOTSUP
+ * with exception to mount and unmount which are assigned NULL.
+ *
+ * @param name name of file_system_t object
+ * @param ... list of designated member initializers of file_system_t
+ */
+#define FILE_SYSTEM_DEFINE(name, ...)					\
+	static struct fs_file_system_t name = {				\
+		.open = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),		\
+		.read = FILE_SYSTEM_UNSUPPORTED_OPERATION(ssize_t),	\
+		.write = FILE_SYSTEM_UNSUPPORTED_OPERATION(ssize_t),	\
+		.lseek = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.tell = FILE_SYSTEM_UNSUPPORTED_OPERATION(off_t),	\
+		.truncate = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.sync = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),		\
+		.close = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.opendir = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.readdir = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.closedir = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.mount = NULL,						\
+		.unmount = NULL,					\
+		.unlink = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.rename = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.mkdir = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		.stat = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),		\
+		.statvfs = FILE_SYSTEM_UNSUPPORTED_OPERATION(int),	\
+		__VA_ARGS__						\
+	}
+
+/**
+ * @}
+ */
+
+#ifdef __cplusplus
```

Comment:
Is the FILE_SYSTEM_DEFINE() macro meaningful or useful in C++? I thought C++ struct  initializers needed to list members in declaration order, so it doesn't look like the `__VA_ARGS__` above would be able to override the defaults in the above definition. Why bother to declare it in `#ifdef __cplusplus`? (Not a C++ expert)

Reply:
Yes, this would not work if somebody implemented an external file system in a C++ source file.  It's not an issue until somebody does that, which is why I'm willing to ignore it.

Reply:
Sorry, I tend to forget we already have C++ since 1983.

Reply:
Will fix it somehow.

Reply:
It'd be nice if we could use C++ in Zephyr, because we could provide default values in the structure definition, then the `fs_file_system` object definition could provide only the ones that need to be replaced, as is done today, without any macro magic.

But without that I don't think there's anything to fix here: it's fine to have a macro that will only work in C99-C18 code.  There's still some like that in Bluetooth, IIRC.

Reply:
> 
> But without that I don't think there's anything to fix here: it's fine to have a macro that will only work in C99-C18 code. There's still some like that in Bluetooth, IIRC.

I am not sure I want it to go through if I fail to fix it for C++. This would force  C++ users to initialize the  structure directly, which would make using the macro pointless. if their code would suppose to compile in both C and C++.

Reply:
> But without that I don't think there's anything to fix here: it's fine to have a macro that will only work in C99-C18 code.

I agree; I just thought it was weird to put this macro definition inside a c linkage block when C++ is used. That usually is a hint to the programmer that the definitions inside work in C++ too. I know this is a macro and therefore has no linkage, but just thought I'd ask.

Not a blocker anyway and marking resolved.

Reply:
> I just thought it was weird to put this macro definition inside a c linkage block when C++ is used. That usually is a hint to the programmer that the definitions inside work in C++ too.

It's the function declarations that are generated by expanding `FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION` that warrant the linkage specification, not the macro definition.  (Though I'd rather have that boilerplate used in all headers, since it reduces the likelihood of adding a declarations and forgetting to also add the linkage specification.)

But looking at that those are a problem, because we'll have copies of them in every file that expands `FILE_SYSTEM_DEFINE()` and uses them.  Better to just declare them here, and have a single definition in some common file that all the `fs_file_system_t` objects can reference.

Unresolved this thread so that can be taken into account when @de-nordic decides how to proceed.

Reply:
> It's the function declarations that are generated by expanding `FILE_SYSTEM_DEFINE_UNSUPPORTED_OPERATION` that warrant the linkage specification, not the macro definition.

I know, which is why my very next sentence was:

> I know this is a macro and therefore has no linkage, but just thought I'd ask.